src/Queue.h

```diff
@@ -128,58 +128,48 @@ template<size_t PARAMETER_SIZE>
 void
 Queue<PARAMETER_SIZE>::put(const Request<PARAMETER_SIZE>& request)
 {
-    m_mutex.lock();
-    if(m_queue.size() >= m_max_elements) {
-        std::cerr << "Message loss in " << m_queue_name << " - queue is full, " << m_max_elements << " allowed"
-                  << std::endl;
-        m_mutex.unlock();
-    } else {
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
+
+        if(is_full()) {
+            return;
+        }
+
         m_queue.push(request);
-        m_mutex.unlock();
-        m_cv.notify_one();
     }
+    
+    m_condition_variable.notify_one();
 }
 
 template<size_t PARAMETER_SIZE>
 void
 Queue<PARAMETER_SIZE>::put(const asn1SccPID sender_pid, const uint8_t* data, size_t length)
 {
-    m_mutex.lock();
-    static Request<PARAMETER_SIZE> request;
-
-    if(length > PARAMETER_SIZE) {
-        std::cerr << "Internal error in " << m_queue_name << " - queue accepts messages with size " << PARAMETER_SIZE
-                  << std::endl;
-        m_mutex.unlock();
-    }
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
 
-    memcpy(request.data(), data, length);
-    request.set_length(length);
-    request.set_sender_pid(sender_pid);
+        if(is_full()) {
+            return;
+        }
 
-    if(m_queue.size() >= m_max_elements) {
-        std::cerr << "Message loss in " << m_queue_name << " - queue is full, " << m_max_elements << " allowed"
-                  << std::endl;
-        m_mutex.unlock();
-    } else {
-        m_queue.push(request);
-        m_mutex.unlock();
-        m_cv.notify_one();
+        m_queue.emplace(sender_pid, data, length);
     }
+
+    m_condition_variable.notify_one();
```

Comment:
I think that manual mutex lock/unlock was used for a reason. The mutex shall be unlocked before notify of condition variable.

Reply:
You are right, and this behavior is preserved.

Reply:
Ok, separate block for mutex.